### PR TITLE
Fix java 17 compilation

### PR DESCRIPTION
**Pull request by keynmol**
This is currently working on top of #76 

1. pass --add-exports flags to javac when we're on Java 17
   - This has corner cases because we need to manage javaHome, both explicitly set and inherited from the build. Setting java home explicitly is required to prevent Zinc from creating a in-process Java compiler that we cannot configure with -J flags.

### Test plan
- Existing tests were modified to exercise the corner case
- Modify CI to run on Java 8, 11, 17

### CODE REVIEW

File: src/main/scala/com/sourcegraph/sbtsourcegraph/SourcegraphEnable.scala [hunk 1]
```diff
@@ -31,12 +31,20 @@ object SourcegraphEnable {
...
       )
       enableSemanticdbPlugin =
         List(
+          Option(
+            javacOptions.in(p) ++= {
+              if (Versions.isJavaAtLeast(17, home = javaHome.in(p).value))
+                javacModuleOptions
+              else Nil
+            }
+          ),
```
Comment (keynmol) [inline src/main/scala/com/sourcegraph/sbtsourcegraph/SourcegraphEnable.scala:47]: On JDK 17+ pass the required parameters to JVM that launches javac.

File: src/main/scala/com/sourcegraph/sbtsourcegraph/SourcegraphEnable.scala [hunk 2]
```diff
@@ -143,4 +164,34 @@ object SourcegraphEnable {
...
+
+  private def calculateJavaHome = {
+    // We can safely use java.home property
+    // on JDK 17+ as it won't be pointing to JRE which
+    // doesn't contain a compiler.
+    if (Versions.isJavaAtLeast(17)) {
+      // On JDK 17+ we need to explicitly fork the compiler
+      // so that we can set the necessary JVM options to access
+      // jdk.compiler module
+      Some(new File(System.getProperty("java.home")))
```
Comment (olafurpg) [inline src/main/scala/com/sourcegraph/sbtsourcegraph/SourcegraphEnable.scala:190]: I'm not 100% sure, but forking javac may cause some builds to fail. Not a blocking comment, just good to keep an eye for it. If we can't turn on forking then we could try to automatically add the exports to SBT_OPTS

File: src/main/scala/com/sourcegraph/sbtsourcegraph/Versions.scala [hunk 3]
```diff
@@ -65,6 +66,67 @@ object Versions {
...
+            case Some(javaHome) =>
+              val sb = new StringBuilder
+              val proc = {
+                val cmd =
+                  if (scala.util.Properties.isWin)
+                    Paths.get("bin", "java")
+                  else Paths.get("bin", "java")
+
+                scala.sys.process
+                  .Process(Seq(cmd.toString(), "-version"), cwd = javaHome)
```
Comment (olafurpg) [inline src/main/scala/com/sourcegraph/sbtsourcegraph/Versions.scala:88]: This is always going to be quite brittle. Ideally, we use something like `SystemJavaVersion` from scip-java. Not a blocking comment, just an FYI if you don't want too maintain this parser forever https://sourcegraph.com/github.com/sourcegraph/scip-java/-/blob/scip-java/src/main/scala/com/sourcegraph/scip_java/buildtools/SystemJavaVersion.scala?L16

**Review by olafurpg — APPROVED** [review]

File: src/main/scala/com/sourcegraph/sbtsourcegraph/SourcegraphEnable.scala (hunk 2 above)
Comment (keynmol) [reply src/main/scala/com/sourcegraph/sbtsourcegraph/SourcegraphEnable.scala:190]: Can you expand on what might be causing issues?

Perhaps I could add an extra test for this just to be sure

Comment (olafurpg) [reply src/main/scala/com/sourcegraph/sbtsourcegraph/SourcegraphEnable.scala:190]: I can't remember any concrete examples. Fine to try this out and just keep an eye out for regressions.

File: src/main/scala/com/sourcegraph/sbtsourcegraph/Versions.scala (hunk 3 above)
Comment (olafurpg) [reply src/main/scala/com/sourcegraph/sbtsourcegraph/Versions.scala:88]: That sounds good.
